homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+
+            if len(event_data['images']) > 0:
+                thumb_index = math.floor(len(event_data['images'])/2)
+                event_data['thumbUrl'] = (
+                    event_data['images'][thumb_index]['url'])
+
+            events_data.append(event_data)
+            count += 1
+
+        return events_data
+
+    def record_stream(self):
+        """ Records individual frames to disk for a period of time """
+        if self.is_recording:
+            return
+        else:
+            self._is_recording = True
+
+        self.update_ha_state(True)
+        self.send_recording_state()
+
+        rec_dir = (
+            'recording-' +
+            datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S'))
+
+        rec_dir_full = os.path.join(self.recording_images_path, rec_dir)
+
+        if not os.path.isdir(rec_dir_full):
+            os.makedirs(rec_dir_full)
+
+        recording_length = DEFAULT_RECORDING_SECONDS
+        recording_started = datetime.datetime.now()
+        recording_time = 0
+        while (
+                recording_time < recording_length
+                and self._is_recording):
+
+            new_file_name = (
+                'recording_image-' +
+                datetime.datetime.now().strftime(
+                    '%Y-%m-%d_%H-%M-%S-%f') + '.jpg')
+
+            new_file_path = os.path.join(rec_dir_full, new_file_name)
+
+            response = self.get_camera_image()
+            open(new_file_path, 'wb').write(response.content)
+
+            recording_time = (
+                datetime.datetime.now() - recording_started).total_seconds()
+
+        self._is_recording = False
+        self.update_ha_state(True)
+        self.send_recording_state()
+
+    def stop_recording(self):
+        """ Stop recording camera stream """
+        self._is_recording = False
+
+    def take_snapshot(self):
+        """ Records individual frames to disk for a period of time """
+        if self.is_taking_snapshot:
+            return
+        else:
+            self._is_taking_snapshot = True
+
+        self.send_snapshot_state()
+
+        time_str = datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S-%f')
+        rec_dir = ('recording-' + time_str)
+
+        rec_dir_full = os.path.join(self.snapshot_images_path, rec_dir)
+
+        if not os.path.isdir(rec_dir_full):
+            os.makedirs(rec_dir_full)
+
+        new_file_name = ('recording_image-' + time_str + '.jpg')
+
+        new_file_path = os.path.join(rec_dir_full, new_file_name)
+
+        response = self.get_camera_image()
+        open(new_file_path, 'wb').write(response.content)
+
+        self._is_taking_snapshot = False
+        self.update_ha_state()
+        self.send_snapshot_state()
+
+    @property
+    def images_path(self):
+        """ The base path for the location of images such as snapshots,
+            recordings and motion capture events """
+        if self._images_path is None:
+            default_images_path = os.path.join(
```

Comment:
I put it in [Pivotal](https://www.pivotaltracker.com/story/show/94258262) to allow `hass.config.path` to take multiple arguments so you can do
```
self.hass.config.path(CAMERA_CONFIG_PATH, self.entity_id)
```